### DISCUSSION TITLE
[Feature] Simple notes navigation

### DISCUSSION THREAD

**Discussion by mjthoraval** [windingwind/zotero-better-notes#439] [Ideas]
I am very confused by the current notes navigation options.
If I click on a zotero://note link, it can either open in a new window, in the same pane, in the right side pane or in the workspace, depending if the note where you click the link is opened in a separate window, in a right side pane next to the library or a PDF Viewer tab, in the workspace or in the right side of the workspace. It can even depend on whether the workspace is currently opened as a tab in the main Zotero window or not.

I am mostly taking notes in Evernote, where the behaviour of links is straightforward:
- I copy an internal link from Right click on the note -> Copy Internal Link (or Ctrl+Alt+L) -> Paste the link into the target note.
- I can create an internal link between notes simply by drag and drop a note into another.
- A single click on the internal link directly takes me to the target note in the same pane.
- I can open the target note in a separate window from Shift+Click.

I am not interested in complex bi-directional links, workspace or split view, before the most simple notes navigation can work efficiently. Is it possible to get the very simple notes navigation features described above?
I guess most note taking apps follow similar conventions. Are you following the logic of some other note taking app to build the current behaviour?

**Comment by windingwind** [comment]
Sorry for the possible misleading of the opening note link behavior. Currently, it behaves like this:
- If there is a workspace window/tab alive:
  - Focus the workspace and open the note on the right side (known as the preview pane). If it is the main note, it should have been displayed in the center of the workspace (known as the main note pane) so we do not open it twice.
- If no workspace is open:
  - Open in a standalone window.

The note pane on the right side of the library tab is a preview provided by Zotero and I think we may better keep it as is.
The note pane on the PDF viewer tab is also a Zotero feature. 

It is possible to let the plugin take over all opening note cases, but disabling Zotero features would really "have fundamentally changed the functionalities of Zotero".

Let's see what we already have and what I can do to improve your experience:

> I copy an internal link from Right click on the note -> Copy Internal Link (or Ctrl+Alt+L) -> Paste the link into the target note.

You could copy links here:

![image](https://user-images.githubusercontent.com/33902321/194757764-f4fccb7f-71d4-4f73-85f7-ebf408e80864.png)
 
I could try to provide some shortcuts for these functions.

> I can create an internal link between notes simply by drag and drop a note into another.

The drag&drop between notes is controlled by Zotero. The dragged note content will be put into the drop position. That's what we better not touch.

> A single click on the internal link directly takes me to the target note in the same pane.

This is a behavior of opening note links in the very early version of the plugin. Users complained that clicking the note links should act the same as other links, so I removed that. Maybe we could make a switch for users.

> I can open the target note in a separate window from Shift+Click.

That's reasonable.

Would it be acceptable after the changes above?

**Comment by mjthoraval** [comment]
It is nice indeed to leave as much as possible the Zotero features untouched. So I will try to propose a few changes in line with this idea:

1. Option to open zotero://note/ links from a single click vs. the 2 clicks with the preview for other types of Zotero links. Keep it as an option if others prefer the longer opening process. But it is a bit weird, as all other note taking apps I have tested actually open internal links with a single click.

2. I would also consider adding a separate option to also extend this behaviour to the zotero://select/ and the zotero://open-pdf links. But as you mention, you may need to be more careful with this, as it is the way it works by default in Zotero. But I think the logic behind the current status is:
- The behaviour of the select and open-pdf Zotero links simply follow the format of the external http links.
- I guess the reason to display the external links before opening is for security reasons: you want to make sure where the link will take you to before opening it.
- In this logic, it is actually not necessary to display the internal links in the same way as external links, because there is no security risk associated with internal links.
- The key limitation for Zotero to do anything about these links is that they are actually not supposed to be used directly in the notes at the moment (any counter example?). All the features to make the links available in Zotero include a more complex citation structure with a popup coming out to get the different links. As they are not supposed to be used anyway, there is no reason for Zotero to change the way they behave.
In this logic, claiming that we should keep the same behaviour as the other links does not seem to be a valid argument I think, as the current situation is probably just because Zotero did not really care about links that should not be used anyway. The main considerations here should really be:
- Is this going to break something in Zotero?
- If there is no technical constrain, what do the users actually prefer?
I could also raise a request in the Zotero Forums about this. But this would work only if I can demonstrate that these links are actually used in a logical and efficient way for a specific workflow, which I cannot find at the moment.

3. Trigger the preview and options with Right Click on the note links instead of the current Left Click.

4. Option to display (or not) the note preview when hovering over the link.

5. Open the zotero://note/ links into a new window with Shift+Click.

6. Directly open a zotero://note/ link into the same pane/window instead of opening a new window, for all configurations outside of the Workspace. I am not sure yet how to use the Workspace, so I could check again later if I start using it. This is a plugin only link, so it will not affect the core Zotero features. This should probably be the default if there is no complain from other users, as the opening in a new window is usually triggered by Shift+Click.

7. Together with the first point, I would like to have the option to change the colour of the internal links zotero://note/ to green (colour may be optional also), so that I can directly identify them vs the other existing types of Zotero links. Hopefully that colouring can remain after disabling the Better Notes plugin, so that it also helps me to identify the broken links if I decide the stop using the plugin. I see that other users may prefer to follow other rules for colouring the links. But I still believe that giving a default green colour for the zotero://note/ links (if not overridden by other options) would be the most beneficial for most users.

8. Option to hide the "Open Workspace" above "My Library". I may experiment with it later, but I want to avoid using it at the moment.

9. Add the options to Copy Note Link (zotero://note/...) when a note (standalone note or child note) is selected in the main library tab:
- From a right click on the item note or child note.
- With a shortcut Ctrl+Alt+L when a note is selected.

10. Add the options to Copy Note Link of Current Line when inside a note:
- From right click on a line inside the note. In this way, no need to go all the way up to the top of the note anymore. The link can be copied quickly directly from the line you want to link to.
- The same shortcut Ctrl+Alt+L could also be used to Copy Note of Current Line when used inside a note. The link to the note would still be available from the top right 3 dots menu, so it probably does not need to make another shortcut for this.

11. I would actually also extend the Copy Note Link behaviour to a Copy PDF Link (zotero://open-pdf/...) when an item with a pdf file is selected in the library, or also directly on any child item PDF file. I don't think it is currently possible to get that open-pdf link in any other way?

12. I think the current  drag & drop behaviour of Zotero with notes does not make any sense at the moment. I don't understand why anyone would like to duplicate the content of a note in a way that does not synchronize with the original note. It probably behaves like this because the internal links between notes are not working yet. But this is probably a secondary issue that can be addressed later when internal links are included in Zotero.

Feel free to reject any of the requests above as you think they are too complicated or modify too much Zotero.
The main point of this post is that internal links between notes is a very powerful tool to enhance the note taking in Zotero.
I think that they have a good chance to be integrated to Zotero in the future (hopefully).
But the Workspace and Main Note features are another level of complexity. More complexity is less general, so less likely to be integrated into Zotero.
So I would recommend making the internal links fully functional without the Workspace and Main Note features before considering how they interact with them.

Thank you very much again for your hard work on this plugin and the amazing features it brings to Zotero!

**Comment by windingwind** [comment]
My conclusions of current stage:

> Option to open zotero://note/ links from a single click
> Open other links in the same way
> Trigger the preview and options with Right Click on the note links instead of the current Left Click.
> Option to display (or not) the note preview when hovering over the link.
> Open the zotero://note/ links into a new window with Shift+Click.

- [x] Left-click(+with shift), right-click(+with shift), hover, cursor-enter, and shortcut. Users can assign different behaviors to them, like opening in the workspace preview pane, setting as the main note (opening in the workspace main note pane), opening in a standalone window, previewing, copying links, etc.

- [ ] Option to hide the "Open Workspace" above "My Library".

- [ ] Add the options to Copy Note Link (zotero://note/...) when a note (standalone note or child note) is selected in the main library tab, and open-pdf or select link. With shortcut Ctrl+Alt+L.
I'll keep the default behaviors consistent with the current ones.

If I missed anything, please let me know.

Q&A
> Is this going to break something in Zotero?

I guess not. The problem is, you may not have much chance to see other types of zotero:// links in the notes.

> If there is no technical constrain, what do the users actually prefer?

We could leave the choice to the users.

> Together with the first point, I would like to have the option to change the colour of the internal links zotero://note/ to green (colour may be optional also), so that I can directly identify them vs the other existing types of Zotero links. Hopefully that colouring can remain after disabling the Better Notes plugin, so that it also helps me to identify the broken links if I decide the stop using the plugin. I see that other users may prefer to follow other rules for colouring the links. But I still believe that giving a default green colour for the zotero://note/ links (if not overridden by other options) would be the most beneficial for most users.

Changing color would be a little bit complex, and impossible if the plugin is disabled.

As we are going to have a totally different prefs window in Zotero 7 and it is currently unavailable in Zotero 6, I try to avoid adding a pref pane before Zotero 7. So the options will be put in a note template in JSON format.

**Comment by mjthoraval** [comment]
Yes, that would be great. Thank you very much.

I will ask in the Zotero Forums if there is a way to change the colour of certain links.
Hopefully Zotero can add support for that and can be added to the plugin later.
Here is the link to the post in the Forums (no reply yet): https://forums.zotero.org/discussion/100302/change-font-colour-for-some-links

**Comment by windingwind** [comment]
@mjthoraval 
Customizing actions of click, click with alt/shift/ctrl in Edit-Preferences-Advanced-Settings Editor-linkAction

![image](https://user-images.githubusercontent.com/33902321/201688054-a26205e2-5a58-4ef9-85d2-45fbc32fb2c5.png)

Available actions are:

openLink, openLinkIfNote, openLinkIfSelect, openLinkIfPDF, openLinkInNewWindow, copyLink, setMainNote

**Reply by mjthoraval** [reply]
Would it be possible to recover this feature?
Nearly all the note editors can open a link from a single click:
https://forums.zotero.org/discussion/112220/zotero-7-beta-single-click-to-open-links
